src/storage/exec/UpdateNode.h

```diff
@@ -167,57 +167,65 @@ class UpdateTagNode : public UpdateNode<VertexID> {
-                        this->exeResult_ = this->collTagProp(vId);
-                    } else {
-                        this->exeResult_ = kvstore::ResultCode::ERR_KEY_NOT_FOUND;
-                    }
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            return ret;
+        }
 
-                    if (this->exeResult_ != kvstore::ResultCode::SUCCEEDED) {
-                        return folly::none;
-                    }
-                    return this->updateAndWriteBack(partId, vId);
-                } else {
-                    // if tagnode/edgenode error
-                    return folly::none;
-                }
-            },
-            [&ret, &baton, this] (kvstore::ResultCode code) {
-                if (code == kvstore::ResultCode::ERR_ATOMIC_OP_FAILED &&
-                    this->exeResult_ != kvstore::ResultCode::SUCCEEDED) {
-                    ret = this->exeResult_;
-                } else {
-                    ret = code;
-                }
-                baton.post();
-            });
+        if (this->planContext_->resultStat_ == ResultStatus::ILLEGAL_DATA) {
+            return kvstore::ResultCode::ERR_INVALID_DATA;
+        } else if (this->planContext_->resultStat_ ==  ResultStatus::FILTER_OUT) {
+            return kvstore::ResultCode::ERR_RESULT_FILTERED;
+        }
+
+        if (filterNode_->valid()) {
+            this->reader_ = filterNode_->reader();
+        }
+        // reset StorageExpressionContext reader_, because it contains old value
+        this->expCtx_->reset();
+
+        if (!this->reader_ && this->insertable_) {
+            ret = this->insertTagProps(partId, vId);
+        } else if (this->reader_) {
+            this->key_ = filterNode_->key().str();
+            ret = this->collTagProp(vId);
+        } else {
+            ret = kvstore::ResultCode::ERR_KEY_NOT_FOUND;
+        }
+
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            return ret;
+        }
+
+        auto batch = this->updateAndWriteBack(partId, vId);
+        if (batch == folly::none) {
+            return kvstore::ResultCode::ERR_INVALID_DATA;
+        }
+
+        std::vector<VMLI> dummyLock = {std::make_tuple(planContext_->spaceId_,
+                                                       partId, tagId_, vId)};
+        nebula::MemoryLockGuard<VMLI> lg(planContext_->env_->verticesML_.get(),
+                                         std::move(dummyLock));
+        if (!lg) {
+            auto conflict = lg.conflictKey();
+            LOG(ERROR) << "vertex conflict "
+                       << std::get<0>(conflict) << ":"
+                       << std::get<1>(conflict) << ":"
+                       << std::get<2>(conflict) << ":"
+                       << std::get<3>(conflict);
+            return kvstore::ResultCode::ERR_DATA_CONFLICT_ERROR;
+        }
+
+        folly::Baton<true, std::atomic> baton;
+        auto callback = [&ret, &baton] (kvstore::ResultCode code) {
```

Comment:
Please put `lg` in the callback function to avoid inconsistent life cycle.

Reply:
> Please put `lg` in the callback function to avoid inconsistent life cycle.

Aha, don't need put 'lg' to callback at here. `folly::Baton` will control the lifecycle.

Reply:
yeah. `baton.wait()`  has waited.